node/cn/peer.go

```diff
@@ -965,30 +965,42 @@ func (p *multiChannelPeer) UpdateRWImplementationVersion() {
 
 func (p *multiChannelPeer) ReadMsg(rw p2p.MsgReadWriter, connectionOrder int, errCh chan<- error, wg *sync.WaitGroup, closed <-chan struct{}) {
 	defer wg.Done()
-	for {
-		msg, err := rw.ReadMsg()
-		if err != nil {
-			p.GetP2PPeer().Log().Warn("ProtocolManager failed to read msg", "err", err)
-			errCh <- err
-			return
-		}
 
-		msgCh, err := p.chMgr.GetChannelWithMsgCode(connectionOrder, msg.Code)
-		if err != nil {
-			p.GetP2PPeer().Log().Warn("ProtocolManager failed to get msg channel", "err", err)
-			errCh <- err
-			return
-		}
+	var (
+		readMsgCh     = make(chan p2p.Msg, channelSizePerPeer)
+		readLoopEndCh = make(chan struct{})
+	)
 
-		if msg.Size > ProtocolMaxMsgSize {
-			err := errResp(ErrMsgTooLarge, "%v > %v", msg.Size, ProtocolMaxMsgSize)
-			p.GetP2PPeer().Log().Warn("ProtocolManager over max msg size", "err", err)
-			errCh <- err
-			return
+	go func() {
+		for {
+			msg, err := rw.ReadMsg()
+			if err != nil {
+				p.GetP2PPeer().Log().Warn("ProtocolManager failed to read msg", "err", err)
+				errCh <- err
+				close(readLoopEndCh)
```

Comment:
```suggestion
```
@yoomee1313, Good to know a read timeout was already placed there. Okay looks fine. Another concern from here is closing the closed channel. The possible scenario is that `readLoopEndCh` may be already closed, but the late closing again by the timeout can encounter runtime panic. Removal of this code looks fine in my view. if the right above line `errCh <- err` is executed, the loop (line 987 ~ 1007) will be exited itself by line 1111.

Reply:
@hyunsooda yes, you're right. I removed the chennel. Thanks!